### PR TITLE
Update validateUILayouts logic

### PR DESCRIPTION
**Pull request by denisonbarbosa**
We used to have a static verification for the UI layouts returned by the broker, but those are subject to change according to what the system can currently handle. This means that it's up to PAM to provide which layouts are supported, which fields of those layouts are required and what values those fields support.

This replaces the previous static implementation of validateUILayouts with a dynamic one that takes into consideration what was specified by PAM.

UDENG-1166

### CODE REVIEW

File: internal/brokers/broker.go
Comment (denisonbarbosa) [inline internal/brokers/broker.go:226]: Ok, so this was a design decision that I'm not totally 100% sure of, but I thought it was worth pushing it so we can talk about it...

The thing is: we used to always assign everything to the returned layout even if it was an empty value (see the changes in the golden files). I'm not sure that was the best approach, tbh. 
I think it's best to validate and return only what was supposed to be there (i.e. required files or optional fields that were set) instead of returning everything. This means it will be up to the UI renderer to check if the values are there or not before rendering them and I think this is the best approach, since we end up preventing content injections that could come from the brokers.

wdyt?

File: internal/brokers/broker.go
Comment (denisonbarbosa) [inline internal/brokers/broker.go:100]: Should this be called every time we call GetAuthenticationModes or just when we don't have the validators (i.e. `b.layoutValidators == nil`)? There is a compelling case for both options, tbh.

Comment (didrocks) [reply internal/brokers/broker.go:100]: right, as we will go over this loop multiple times in the future for multiple MFA, we may have different requirements (first one without browser support, then browser is supported for instance), so it’s safest to do it everytime we return (from the PAM perspective) the constraints.

**Review by didrocks — CHANGES_REQUESTED** [review]
I really like this one, it’s clean (I have some suggestions/requests for more thoughts on the naming), but the validation function is really what I was thinking about.

I think the testing though needs to be digged in: having more use cases and, in particular, more precise one, not linking with the layout we support, because apart from "type" (which is one of the case to tests, the rest is free form, and we can reflect what we are testing in the input name.

File: internal/brokers/broker.go
Comment (denisonbarbosa) [reply internal/brokers/broker.go:200]: You can't even imagine how much I suffered with naming stuff in this PR :joy:

File: internal/brokers/broker.go
Comment (denisonbarbosa) [reply internal/brokers/broker.go:191]: I agree, but I also don't really like the "attach" variant as it kind of makes you think that the validators are already there somewhere and we only add them to the broker, no? Maybe we should just remove the verb entirely and go for `validators()` :joy:

Comment (didrocks) [reply internal/brokers/broker.go:191]: we need to still convery that we are having a side effect on the broker, or… maybe just return the validators and the caller will be `b.validators = generateValidators(…)` or something along the like?

File: internal/brokers/broker_test.go
Comment (denisonbarbosa) [reply internal/brokers/broker_test.go:165]: I agree, so I'll wait until we merge the service tests before updating the tests and then we take a look at it.

Comment (didrocks) [reply internal/brokers/broker_test.go]: So, you want to put that one on hold until we get pr 22 merged and rebase on it?

Basically, I would like that one to be merged with the proper, descriptive tests, or we will miss some use cases.

File: internal/brokers/broker.go
Comment (didrocks) [reply internal/brokers/broker.go:217]: It’s a programming error from the client side (client being PAM/gdm/…). I think it’s fine then to say, as we are in control of it, that we don’t trust the input anymore as long as we log the error.

File: internal/brokers/broker.go
Comment (didrocks) [reply internal/brokers/broker.go:182]: I don’t see the comment with the struct comment as you did on MM yesterday?

**Comment by didrocks** [comment]
Ok, so I think you missed 2 comments that are addressable right now, and if I get you right, you prefer to have pr 22 landed before doing the test changes (even if I don’t forsee any conflicts TBH, looking at both), correct?

File: internal/brokers/broker.go
Comment (denisonbarbosa) [reply internal/brokers/broker.go:217]: I've rewritten it to return an error if the type is invalid. Wdyt? Either way is fine for me, tbh.

File: internal/brokers/broker.go
Comment (denisonbarbosa) [reply internal/brokers/broker.go:182]: It's in the diff for me, right below this comment as the matter of fact :joy:

Comment (didrocks) [reply internal/brokers/broker.go:182]: either I had no coffee, either GH played with me! You are correct, it’s here :)

File: internal/brokers/broker_test.go
Comment (didrocks) [reply internal/brokers/broker_test.go]: Agreed, I would go for this one which is closer to resolution! :)

File: internal/brokers/broker.go
Comment (denisonbarbosa) [reply internal/brokers/broker.go:217]: Sounds better, IMHO. We log it and then ignore the layout that's invalid.

File: internal/brokers/broker.go
Comment (didrocks) [inline internal/brokers/broker.go:194]: hum, we are missing some context in the log, which session ID this has been provided? I think then we have to bubble the error up and handling logging in the caller. However, as said in the previous one, it means that theorically, we should not trust the fieldValidator returned value.

So… maybe pass the sessionID just for the sake of logging? Not ideal, but I don’t have a better idea atm :/

File: internal/testutils/broker.go
Comment (didrocks) [inline internal/testutils/broker.go:173]: is it nil or an empty map? (or should we have both?)

File: internal/brokers/broker_test.go
Comment (didrocks) [inline internal/brokers/broker_test.go:21]: you have spaces here, we don’t really return this default, so maybe trim that? If we want to assert one the support, this is an additional test case (and so layout) maybe?

File: internal/brokers/broker_test.go
Comment (didrocks) [inline internal/brokers/broker_test.go:118]: you meant in the PAM module?

File: internal/brokers/testdata/TestGetAuthenticationModes/golden/get_authentication_modes_and_generate_validators
Comment (didrocks) [inline internal/brokers/testdata/TestGetAuthenticationModes/golden/get_authentication_modes_and_generate_validators:2]: I think here we can have one test case returning more than one mode. (needs changes in the broker). This aligns with my comment on the internal/services/pam test branch.

File: internal/brokers/export_test.go
Comment (didrocks) [inline internal/brokers/export_test.go:36]: a small, nice comment on top for our next future selves? :p

**Review by didrocks — CHANGES_REQUESTED** [review]
excellent! As said, I love the fixes and in particular the new mock test approach which is independent of the layout type. I have some minor comments that should be easy to address!

File: internal/brokers/broker_test.go
Comment (denisonbarbosa) [reply internal/brokers/broker_test.go:118]: What do you mean? I meant the validator's generation. They are generated in the GAM method, but since here we don't call it, we need to expose the function that generates the validators and call it manually.

File: internal/brokers/broker.go
Comment (denisonbarbosa) [reply internal/brokers/broker.go:194]: We could attach the sessionID to the context we pass for this function, wdyt?

Comment (didrocks) [reply internal/brokers/broker.go:194]: That’s one way I was thinking about too. Unfortunately, it means that we are going to again have to pass the context everywhere we want to log in. As we are not as deep as in adsys, I think passing the value is the most straightforward way for now. Let’s revisit if we get this case in more places if you don’t mind.

File: internal/testutils/broker.go
Comment (didrocks) [reply internal/testutils/broker.go:173]: … for now, so that’s why a dedicated tests can prevent us from regressing?

File: internal/brokers/broker_test.go
Comment (didrocks) [reply internal/brokers/broker_test.go]: Yes, I didn't think about it but trimming after the split is easy enough to do and be more permissive. Then, having a test for it (I think it’s small enough to get into that PR)

Comment (denisonbarbosa) [reply internal/brokers/broker_test.go]: Do we want a test function just for it? Or should we attach it to the GAM tests (it could be a simple test case)?

Comment (didrocks) [reply internal/brokers/broker_test.go]: oh testcase, definitively.

File: internal/brokers/broker_test.go
Comment (didrocks) [inline internal/brokers/broker_test.go:103]: That was almost finale :p I would say ignoring whitespaces in accepted values or so. (we don’t allow whitespace around required/optional)

**Review by didrocks — APPROVED** [review]
One the test case being clarified, feel free to stash and merge! This is way better :)

**Pull request merged**
